tools/pkg/solver/registry-solver.toit

```diff
@@ -0,0 +1,260 @@
+
+  hash-code -> int:
+    return url.hash-code
+
+
+class PartialSolution:
+  partial-packages/Map  // PackageDependency -> PartialPackageSolution.
+  unsolved-packages/Deque  // A queue of PackageDependencies that have unresolved partial solutions.
+  url-to-dependencies/Map := {:}  // string -> [PackageDependency], keeping track of the same url with different constraints.
+  solver/Solver
+
+  constructor .solver .partial-packages/Map:
+    unsolved-packages = Deque.from partial-packages.keys
+    partial-packages.keys.do: | dependency/PackageDependency |
+      append-to-list-value url-to-dependencies dependency.url dependency
+
+  /** Performs a deep copy to support backtracking. */
+  constructor.copy other/PartialSolution:
+    solver = other.solver
+    url-to-dependencies = other.url-to-dependencies.map: | _ v | v.copy
+    unsolved-packages = Deque.from other.unsolved-packages
+    package-translator := IdentityMap  // Mapping old PartialPackageSolution's to copied version.
+    partial-packages = copy-dependency-to-solution-map_ other.partial-packages package-translator
+
+  is-solution -> bool:
+    return unsolved-packages.is-empty
+
+  add-partial-package-solution
+      dependency/PackageDependency
+      package/PartialPackageSolution
+      new-package/PartialPackageSolution:
+    package.dependencies[dependency] = new-package
+    partial-packages[dependency] = new-package
+    add-to-set-value url-to-dependencies dependency.url dependency
+    unsolved-packages.add dependency
+
+  refine -> PartialSolution?:
+    if is-solution: return this
+
+    unsolved-dependency/PackageDependency := unsolved-packages.remove-first
+    unsolved-package/PartialPackageSolution := partial-packages[unsolved-dependency]
+
+    package-versions/List := unsolved-package.versions
+
+    package-versions.do: | next-version/SemanticVersion |
+      copy := PartialSolution.copy this
+      if copy.load-dependencies unsolved-dependency next-version:
+        if refined := copy.refine: return refined
+
+    return null
+
+  load-dependencies unresolved-dependency/PackageDependency next-version/SemanticVersion -> bool:
+    description := solver.registries.retrieve-description unresolved-dependency.url next-version
+
+    if not description.satisfies-sdk-version solver.sdk-version:
+      return false
+
+    package/PartialPackageSolution := partial-packages[unresolved-dependency]
+    package.description = description
+    package.versions = null
+
+    description.dependencies.do: | dependency/PackageDependency |
+      if url-to-dependencies.contains dependency.url:
+        partial-package-solutions := IdentitySet
+        url-to-dependencies[dependency.url].do: partial-package-solutions.add partial-packages[it]
+
+        if partial-package := dependency.find-satisfied-package partial-package-solutions:
+          partial-package.add-source-dependency dependency
```

Comment:
```suggestion
          // One partial package has at least one version that satisfies this dependency.
          // Shrink the set of all allowed versions in that partial package so it also takes the current
          // requirement into account. If it is a fixed package then 'add-source-dependency' won't
          // do anything, since the single version is known to work.
          partial-package.add-source-dependency dependency
```